### PR TITLE
[Finishes #148622189] Allow protocol override with HTTPS

### PR DESCRIPTION
**Pull request by jchang2014**
https://www.pivotaltracker.com/story/show/148622189

### CODE REVIEW

File: lib/git_local/repository.rb [hunk 1]
```diff
@@ -104,5 +109,10 @@ def reset_to_latest_from_origin
     def repo_path
       @repo_path ||= "#{local_directory}/#{org_repo}"
     end
+
+    def valid_protocol(protocol)
+      raise InvalidProtocol, "Protocol must be either HTTPS or SSH." unless ["SSH", "HTTPS"].include?(protocol)
```
Comment (erisfully) [inline lib/git_local/repository.rb]: "SSH" and "HTTPS" might be better expressed in a constant like:

```ruby
VALID_PROTOCOLS = {
  ssh: "SSH",
  https: "HTTPS"
}.freeze
```

Then we would use `VALID_PROTOCOLS[:ssh]` etc.

File: lib/git_local/repository.rb [hunk 2]
```diff
@@ -6,15 +6,19 @@ class NotFound < StandardError
     class InvalidArgument < StandardError
     end
 
+    class InvalidProtocol < StandardError
+    end
+
     GITHUB_HOST = "github.com".freeze
 
-    def initialize(org:, repo:, branch:, local_directory:, host: GITHUB_HOST)
+    def initialize(org:, repo:, branch:, local_directory:, host: GITHUB_HOST, protocol: "SSH")
```
Comment (erisfully) [inline lib/git_local/repository.rb]: Same re: constant

File: lib/git_local/repository.rb [hunk 3]
```diff
@@ -81,8 +85,9 @@ def check_for_special_characters(*args)
 
     def clone_and_checkout
       FileUtils.makedirs(repo_path) unless Dir.exist?(repo_path)
+      repo_endpoint = @protocol == "SSH" ? "git@#{host}:#{org_repo}.git" : "https://#{host}/#{org_repo}.git"
```
Comment (erisfully) [inline lib/git_local/repository.rb]: Same re: constant

File: lib/git_local/repository.rb [hunk 4]
```diff
@@ -104,5 +109,10 @@ def reset_to_latest_from_origin
     def repo_path
       @repo_path ||= "#{local_directory}/#{org_repo}"
     end
+
+    def valid_protocol(protocol)
```
Comment (erisfully) [inline lib/git_local/repository.rb]: It feels a little odd to be both validating and returning here. What if this was `validate_protocol`, and we simply did the assignment in the consume case elsewhere?

File: lib/git_local/repository.rb (hunk 3 above)
Comment (erisfully) [reply lib/git_local/repository.rb]: This might be more clear as its own private method `repo_endpoint` (repo_url?)

**Review by erisfully — CHANGES_REQUESTED** [review]

File: lib/git_local/repository.rb (hunk 2 above)
Comment (jchang2014) [reply lib/git_local/repository.rb]: resolved

File: lib/git_local/repository.rb (hunk 3 above)
Comment (jchang2014) [reply lib/git_local/repository.rb]: resolved

File: lib/git_local/repository.rb (hunk 4 above)
Comment (jchang2014) [reply lib/git_local/repository.rb]: resolved

File: lib/git_local/repository.rb (hunk 1 above)
Comment (jchang2014) [reply lib/git_local/repository.rb]: resolved

File: lib/git_local/repository.rb [hunk 5]
```diff
@@ -6,15 +6,23 @@ class NotFound < StandardError
...
+    }.freeze
 
-    def initialize(org:, repo:, branch:, local_directory:, host: GITHUB_HOST)
+    def initialize(org:, repo:, branch:, local_directory:, host: GITHUB_HOST, protocol: VALID_PROTOCOLS[:ssh])
       check_for_special_characters(org, repo, branch, local_directory)
       @branch = branch
       @host = host
       @local_directory = local_directory
       @org = org
       @repo = repo
```
Comment (erisfully) [inline lib/git_local/repository.rb:24]:
```
@protocol = protocol
validate_protocol
```

File: lib/git_local/repository.rb [hunk 6]
```diff
@@ -104,5 +112,14 @@ def reset_to_latest_from_origin
     def repo_path
       @repo_path ||= "#{local_directory}/#{org_repo}"
     end
+
+    def validate_protocol(protocol)
+      raise InvalidProtocol, "Protocol must be either HTTPS or SSH." unless VALID_PROTOCOLS.values.include?(protocol)
+      @protocol = protocol
```
Comment (erisfully) [inline lib/git_local/repository.rb]: Then remove this line this line

File: lib/git_local/repository.rb (hunk 5 above)
Comment (jchang2014) [reply lib/git_local/repository.rb:24]: resolved

**Review by erisfully — APPROVED** [review]

**Pull request merged**